docs/architecture/adrs/005-chat-completion-support.md

```diff
@@ -0,0 +1,74 @@
+# ADR 005: Design for chat completion support
+
+To enable running guardrails on chat completion use-cases, we want to provide a chat completion API that allows users to run various types of detectors in their chat completion flow.
+
+## Motivation
+
+ The [OpenAI chat completions](https://platform.openai.com/docs/guides/chat-completions) feature has been a popular implementation of using generative models to respond in chat conversations. With various conversation use cases such as customer service, we want to expand our API to not only support applying detectors on text in general, but specifically on chat messages.
+
+## Decisions
+
+### General
+- We will use the [OpenAI chat completions API](https://platform.openai.com/docs/api-reference/chat) as our the base API for chat completions. Parameters detailed here can be referenced.
+- Different chat completions server solutions like [vllm](https://docs.vllm.ai/en/stable/index.html) may provide [extra parameters](https://docs.vllm.ai/en/stable/serving/openai_compatible_server.html#extra-parameters-for-chat-api). For the orchestrator API to remain more agnostic of particular server implementation, the orchestrator can pass through any additional parameters to the generation server, which can do validation of parameters.
+- Guardrails-related API updates would be additive, and we will not modify any OpenAI API elements, like `messages` or `choices`.
+    - We will add a `detectors` block in the request, containing `input` and `output` blocks. This will provide flexibility and control, allowing users to provide list of detectors with parameters for both chat completions input and output separately.
+    - Here, `input` and `output` does not strictly refer to whether only the `input` to chat completions is provided to the detectors or that only the `output` to chat completions is provided to the detectors. Detectors specified in `output` can potentially take both input and output of a chat completions model.
+    - We will add a `detections` block in the response.
+- If no input or output detectors were requested, then the `detections.<missing element / input or output>` would not be present in the response.
+- If none of input or output detectors are requested, then we will return `422` (similar behavior to other standalone `v2` endpoints).
+
+
+### Response behavior
+
+#### Assumptions regarding choices
+This is particular to the [`choices` array](https://platform.openai.com/docs/api-reference/chat/object#chat/object-choices) on the chat completion response.
+1. Each choice is independent.
+2. Each choice is generated with the same input.
+
+#### Response
+- If the LLM (model used to generate chat completions) doesn't generate `content` in any `message` in `choices`, then we will ignore the `choices` and return a warning about `detections.output`.
```

Comment:
I think we should only return the warning if `output` is present in the request.

Reply:
> perhaps it just needs to be explicitly written that output detectors are requested?
I think this is what I meant. In a scenario in which only `input` detectors are requested, we probably shouldn't send any warnings regarding `output`.
Does that make sense?

Reply:
I agree this would be implied. I was just unsure on how specific these ADRs need to be. :sweat_smile: